closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java

```diff
@@ -326,40 +328,51 @@ int encrypt(byte[] in, int inOfs, int len, byte[] out, int outOfs) {
      */
     int encryptFinal(byte[] in, int inOfs, int len, byte[] out, int outOfs)
         throws IllegalBlockSizeException, ShortBufferException {
+        byte[] keyCopy;
+        byte[] ivCopy;
+        byte[] aad;
+        int localTagLenBytes;
+        int ret;
+
+        synchronized (this) {
+            localTagLenBytes = tagLenBytes;
+            if (len > (MAX_BUF_SIZE - localTagLenBytes)) {
+                throw new ShortBufferException
+                    ("Can't fit both data and tag into one buffer");
+            }
 
-        if (len > MAX_BUF_SIZE - tagLenBytes) {
-            throw new ShortBufferException
-                ("Can't fit both data and tag into one buffer");
-        }
-
-        if (out.length - outOfs < (len + tagLenBytes)) {
-            throw new ShortBufferException("Output buffer too small");
-        }
+            if ((out.length - outOfs) < (len + localTagLenBytes)) {
+                throw new ShortBufferException("Output buffer too small");
+            }
 
-        checkDataLength(ibuffer_enc.size(), len);
+            checkDataLength(ibuffer_enc.size(), len);
 
-        if (len > 0) {
-            ibuffer_enc.write(in, inOfs, len);
-        }
+            if (len > 0) {
+                ibuffer_enc.write(in, inOfs, len);
+            }
 
-        // refresh 'in' to all buffered-up bytes
-        in = ibuffer_enc.toByteArray();
-        inOfs = 0;
-        len = in.length;
-        ibuffer_enc.reset();
+            // refresh 'in' to all buffered-up bytes
+            in = ibuffer_enc.toByteArray();
+            inOfs = 0;
+            len = in.length;
+            ibuffer_enc.reset();
 
-        byte[] aad = ((aadBuffer == null || aadBuffer.size() == 0) ? emptyAAD : aadBuffer.toByteArray());
+            aad = (((aadBuffer == null) || (aadBuffer.size() == 0)) ? emptyAAD : aadBuffer.toByteArray());
 
-        int ret = nativeCrypto.GCMEncrypt(key, key.length,
-                iv, iv.length,
-                in, inOfs, len,
-                out, outOfs,
-                aad, aad.length, tagLenBytes);
+            // avoid any race conditions in OpenSSL when the key/iv are used concurrently
+            keyCopy = key.clone();
```

Comment:
The clones seem unnecessary, key and iv are already cloned in init(). GCMEncrypt and GCMDecrypt are synchronized now, so they can't run at the same time. I don't see how key/iv can be used concurrently.

Reply:
yeah, that makes sense